File: model_1/CS3233_2023_Midterm/king_s_decree_1_solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    int C;
    cin >> C;
    vector<long long> results;
    for (int c = 0; c < C; c++) {
        int n;
        cin >> n;
        vector<long long> w(n);
        vector<long long> l(n);
        long long total_wealth = 0;
        for (int i = 0; i < n; i++) {
            cin >> w[i];
            total_wealth += w[i];
        }
        for (int i = 0; i < n; i++) {
            cin >> l[i];
        }
        long long total_l = accumulate(l.begin(), l.end(), 0LL);
        sort(l.begin(), l.end());
        vector<long long> distinct;
        vector<int> freqs;
        for (int i = 0; i < n; ) {
            int j = i;
            while (j < n && l[j] == l[i]) {
                j++;
            }
            distinct.push_back(l[i]);
            freqs.push_back(j - i);
            i = j;
        }
        long long cum_freq = 0;
        long long cum_sum = 0;
        long long best = 0;
        int num_d = distinct.size();
        for (int i = 0; i < num_d; i++) {
            long long d = distinct[i];
            int cnt = freqs[i];
            cum_freq += cnt;
            cum_sum += static_cast<long long>(cnt) * d;
            long long numerator = total_wealth - total_l + cum_sum;
            long long candidate_val = numerator / cum_freq;
            long long candidate_seg;
            if (i < num_d - 1) {
                candidate_seg = min(candidate_val, distinct[i+1] - 1);
            } else {
                candidate_seg = candidate_val;
            }
            if (candidate_seg >= d) {
                if (candidate_seg > best) {
                    best = candidate_seg;
                }
            }
        }
        results.push_back(best);
    }
    for (long long res : results) {
        cout << res << '\n';
    }
    return 0;
}

File: model_1/CS3233_2023_Midterm/king_s_decree_1_solution.py
import sys

def main():
    data = sys.stdin.read().split()
    if not data:
        return
    t = int(data[0])
    index = 1
    results = []
    for _ in range(t):
        n = int(data[index])
        index += 1
        w = list(map(int, data[index:index+n]))
        index += n
        l_list = list(map(int, data[index:index+n]))
        index += n
        
        total_wealth = sum(w)
        total_l = sum(l_list)
        l_list.sort()
        distinct = []
        freqs = []
        i = 0
        n_val = len(l_list)
        while i < n_val:
            j = i
            while j < n_val and l_list[j] == l_list[i]:
                j += 1
            distinct.append(l_list[i])
            freqs.append(j - i)
            i = j
        
        cum_freq = 0
        cum_sum = 0
        best = 0
        num_distinct = len(distinct)
        
        for j in range(num_distinct):
            d = distinct[j]
            S_d = total_l + d * cum_freq - cum_sum
            if S_d <= total_wealth:
                if d > best:
                    best = d
            
            count_here = freqs[j]
            cum_freq += count_here
            cum_sum += d * count_here
            
            num = total_wealth - total_l + cum_sum
            candidate_seg = num // cum_freq
            
            if j < num_distinct - 1:
                next_d = distinct[j+1]
                candidate_val = candidate_seg
                if candidate_val > next_d - 1:
                    candidate_val = next_d - 1
            else:
                candidate_val = candidate_seg
                
            if candidate_val > d:
                if candidate_val > best:
                    best = candidate_val
                    
        results.append(str(best))
    
    print("\n".join(results))

if __name__ == '__main__':
    main()